src/Sarif.Sarifer/SpamBackgroundAnalyzer.cs

```diff
@@ -40,7 +41,7 @@ protected override void AnalyzeCore(Uri uri, string text, string solutionDirecto
                 || (this.currentSolutionDirectory?.Equals(solutionDirectory, StringComparison.OrdinalIgnoreCase) != true))
             {
                 // clear older rules
-                this.rules?.Clear();
+                this.rules.Clear();
```

Comment:
>rules [](start = 21, length = 5)

Just curious, why did you make this change? What you've done is to prematurely allocate an object, which will *always* be overwritten in any case. i.e., you create it so that you can call Clear on it and then reassign a new value. Why wouldn't we just perform the null check here?

One reason you might have had in mind is that this could would cause a null deref in, say, cases where a pervious LoadSearchDefinitionsFiles call returns null. If so, I'd suggest a debug assert or an actual verification after that call (that we received a non-null result).

Reply:
imagine the following:
opened a single file (without sln, / spam), that will bring the solutionDirectory to null, since we don't have the path, rules would still be null. With that, when we call the analyze we will have a null reference.
My idea was: to prevent that, i did create the hashset. What I can do is after this, i can check if rules is null, if thats true, i will just return.

---
In reply to: [536222467](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r536222467) [](ancestors = 536222467)